### PR TITLE
GetPublicKey For NCW

### PR DESCRIPTION
**Pull request by goweiss**
Added two new functions for NCW GetPublicKey

### CODE REVIEW

File: src/ncw-api-client.ts [hunk 1]
```diff
@@ -169,4 +173,26 @@ export class NcwApiClient implements NcwSdk {
...
+        let url = `/v1/ncw/${walletId}/public_key_info`;
+        if (args.algorithm) {
+            url += `?algorithm=${args.algorithm}`;
+        }
+        if (args.derivationPath) {
+            url += `&derivationPath=${JSON.stringify(args.derivationPath)}`;
+        }
+        if (args.compressed) {
+            url += `&compressed=${args.compressed}`;
+        }
```
[the remark this answers is not in the document]
Comment (yuval-fireblocks) [reply src/ncw-api-client.ts:187]: also, the code seems duplicated with vault's get pubkey info, lets merge

File: src/ncw-api-client.ts [hunk 2]
```diff
@@ -169,4 +173,26 @@ export class NcwApiClient implements NcwSdk {
...
+        return await this.apiClient.issueGetRequest(url, undefined, { ncw: { walletId } });
+    }
+
+    public async getPublicKeyInfoByAccountAsset(walletId: string, args: PublicKeyInfoByAccountAssetArgs): Promise<PublicKeyResponse> {
```
Comment (yuval-fireblocks) [inline src/ncw-api-client.ts:191]: maybe `getPublicKeyInfoForNCWAsset`, so it's similar to the existing `getPublicKeyInfoForVaultAccount`?

File: src/types.ts [hunk 3]
```diff
@@ -1219,14 +1219,19 @@ export interface PublicKeyInfoArgs {
...
     change: number;
     addressIndex: number;
     compressed?: boolean;
 }
 
+export interface PublicKeyInfoForVaultAccountArgs
+    extends Omit<PublicKeyInfoByAccountAssetArgs, "accountId"> {
+    vaultAccountId: number;
+}
+
```
Comment (yuval-fireblocks) [inline src/types.ts:1234]: omitting accountId for vaultAccoutId, where the base type is `PublicKeyInfoByAccountAssetArgs` is weird.

i think its better to either create a base interface, where ncw\vaults extends it, with the correct fields, or just use one interface, and change the names of the params as needed in each function

Comment (goweiss) [reply src/types.ts:1234]: this is due to backward compatability

File: src/ncw-api-client.ts (hunk 2 above)
Comment (goweiss) [reply src/ncw-api-client.ts:191]: I think `getPublicKeyInfoByAccountAsset` is more meaningful

File: src/types.ts (hunk 3 above)
Comment (yuval-fireblocks) [reply src/types.ts:1234]: there are no backward issues with my suggestion, it's just a better typing style

Comment (yuval-fireblocks) [reply src/types.ts:1234]: something like:

```
export interface BasePublicKeyInfoByAccountAssetArgs {
    assetId: string;
    change: number;
    addressIndex: number;
    compressed?: boolean;
}
export interface PublicKeyInfoForVaultAccountArgs extends BasePublicKeyInfoByAccountAssetArgs {
    vaultAccountId: number;
}
export interface PublicKeyInfoForNCWAccountArgs extends BasePublicKeyInfoByAccountAssetArgs {
    accountId: number;
}
```

File: src/common/public_key_info.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
+import { ApiClient } from "../api-client";
+import { PeerType, PublicKeyInfoArgs, PublicKeyInfoByAccountAssetArgs, PublicKeyInfoForVaultAccountArgs, PublicKeyInformation, PublicKeyResponse, RequestOptions } from "../types";
+import queryString from "query-string";
+
+export async function getPublicKeyInfoImpl(peerType: PeerType, args: PublicKeyInfoArgs, apiClient: ApiClient, walletId?: string): Promise<PublicKeyInformation> {
+    let url: string;
+    let requestOptions: RequestOptions;
+    if (peerType === PeerType.VAULT_ACCOUNT) {
+        url = `/v1/vault/public_key_info`;
+    } else {
```
Comment (yuval-fireblocks) [inline src/common/public_key_info.ts:10]: let's be specific (if (peerType === PeerType.END_USER_WALLET))

File: src/common/public_key_info.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+    return await apiClient.issueGetRequest(url, undefined, requestOptions);
+}
+
+export async function getPublicKeyInfoByAccountAssetImpl(peerType: PeerType, args: PublicKeyInfoForVaultAccountArgs | PublicKeyInfoByAccountAssetArgs, apiClient: ApiClient, walletId?: string): Promise<PublicKeyResponse> {
+    let url: string;
+    let requestOptions: RequestOptions;
+    if (peerType === PeerType.VAULT_ACCOUNT) {
+        url = `/v1/vault/accounts/${(args as PublicKeyInfoForVaultAccountArgs).vaultAccountId}/${args.assetId}/${args.change}/${args.addressIndex}/public_key_info`;
+    } else {
```
Comment (yuval-fireblocks) [inline src/common/public_key_info.ts:30]: same

**Review by YoavBZ — APPROVED** [review]

**Pull request merged**